### PR TITLE
[FLINK-13385][hive]Align Hive data type mapping with FLIP-37

### PR DESCRIPTION
**Pull request by zjuwangg**
## What is the purpose of the change
Align Hive data type mapping with FLIP-37. 

## Brief change log
  - *Align Hive data type mapping with FLIP-37*
  - *Not support interval type mapping for now. It's not type mapping related only, also related to data converting, maybe we can do it in the future.*

## Verifying this change

This change is already covered by existing tests, such as *HiveCatalogDataTypeTest.java*.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): ( no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: ( no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no )
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable )

### CODE REVIEW

**Comment by zjuwangg** [comment]
cc @xuefuz @lirui-apache @twalthr

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveTypeUtil.java [hunk 1]
```diff
@@ -146,6 +143,11 @@ public static TypeInfo toHiveTypeInfo(DataType dataType) {
 			// Flink's primitive types that Hive 2.3.4 doesn't support: Time, TIMESTAMP_WITH_LOCAL_TIME_ZONE
 		}
 
+		if (dataType.equals(DataTypes.BYTES())) {
+			// Hive doesn't support variable-length binary string
```
Comment (xuefuz) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveTypeUtil.java]: The comment doesn't seem applicable in the code context. Let's remove it.

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/table/catalog/hive/HiveCatalogDataTypeTest.java [hunk 2]
```diff
@@ -124,8 +122,8 @@ public void testDataTypes() throws Exception {
 	@Test
 	public void testNonExactlyMatchedDataTypes() throws Exception {
```
Comment (xuefuz) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/table/catalog/hive/HiveCatalogDataTypeTest.java]: Maybe we should change this test to a negative test case, as BINARY and VARBINARY are not supported by hive and it's useful to capture the limitation with a test case.

File: docs/dev/table/catalog.md [hunk 3]
```diff
@@ -179,10 +179,11 @@ Currently `HiveCatalog` supports most Flink data types with the following mappin
...
-| INTERVAL |  N/A |
-| BINARY        |  binary |
-| VARBINARY(p)  |  binary |
-| ARRAY\<E>     |  list\<E> |
+| INTERVAL      |   N/A*** |
+| BINARY        |   N/A |
+| VARBINARY(p)  |   N/A |
+| BYTES         |   BINARY |
+| ARRAY\<E>     |  ARRAY\<E> |
 | MAP<K, V>     |  map<K, V> |
```
Comment (xuefuz) [inline docs/dev/table/catalog.md]: Let's add a note saying that Hive map key type only allows primitive types, while Flink map key can be any data type.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveTypeUtil.java [hunk 4]
```diff
@@ -103,9 +103,6 @@ public static TypeInfo toHiveTypeInfo(DataType dataType) {
 				return TypeInfoFactory.dateTypeInfo;
```
Comment (JingsongLi) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveTypeUtil.java]: Why not use switch case?

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveTypeUtil.java [hunk 5]
```diff
@@ -146,6 +143,11 @@ public static TypeInfo toHiveTypeInfo(DataType dataType) {
 			// Flink's primitive types that Hive 2.3.4 doesn't support: Time, TIMESTAMP_WITH_LOCAL_TIME_ZONE
 		}
 
+		if (dataType.equals(DataTypes.BYTES())) {
```
Comment (JingsongLi) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveTypeUtil.java]: This way will equals nullable, you should put this code in switch case, and check length.

File: docs/dev/table/catalog.md (hunk 3 above)
Comment (JingsongLi) [reply docs/dev/table/catalog.md]: +1

Comment (JingsongLi) [reply docs/dev/table/catalog.md]: Why not throw exception in `HiveTypeUtil.toHiveTypeInfo` when hive not support?

Comment (zjuwangg) [reply docs/dev/table/catalog.md]: > HiveTypeUtil.toHiveTypeInfo

Indeed we throw `UnsupportedOperationException` in `HiveTypeUtil.toHiveTypeInfo`.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveTypeUtil.java (hunk 4 above)
Comment (zjuwangg) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveTypeUtil.java]: good point.

**Comment by zjuwangg** [comment]
cc @JingsongLi @xuefuz to review again.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveTypeUtil.java [hunk 6]
```diff
@@ -85,65 +86,68 @@ public static TypeInfo toHiveTypeInfo(DataType dataType) {
...
-				return TypeInfoFactory.getVarcharTypeInfo(varCharType.getLength());
-			} else if (type.equals(LogicalTypeRoot.DECIMAL)) {
-				DecimalType decimalType = (DecimalType) dataType.getLogicalType();
-
-				// Flink and Hive share the same precision and scale range
-				// Flink already validates the type so we don't need to validate again here
-				return TypeInfoFactory.getDecimalTypeInfo(decimalType.getPrecision(), decimalType.getScale());
+				case VARBINARY: {
+					VarBinaryType varBinaryType = (VarBinaryType) dataType.getLogicalType();
+					if (varBinaryType.getLength() == VarBinaryType.MAX_LENGTH) {
```
Comment (xuefuz) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveTypeUtil.java]: Nit: maybe we can add a comment saying the type is actually DataTypes.BYTES, similar to what is done for VARCHAR.

Comment (zjuwangg) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveTypeUtil.java]: That's an insightful opinion.

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/table/catalog/hive/HiveCatalogDataTypeTest.java [hunk 7]
```diff
@@ -122,20 +121,18 @@ public void testDataTypes() throws Exception {
 	}
 
 	@Test
-	public void testNonExactlyMatchedDataTypes() throws Exception {
+	public void testNonSupportedBinaryDataTypes() throws Exception {
```
Comment (bowenli86) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/table/catalog/hive/HiveCatalogDataTypeTest.java]: we should break this down to two individual tests, one for BINARY and one for VARBINARY

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/table/catalog/hive/HiveCatalogDataTypeTest.java [hunk 8]
```diff
@@ -122,20 +121,18 @@ public void testDataTypes() throws Exception {
...
 		DataType[] types = new DataType[] {
-			DataTypes.BINARY(BinaryType.MAX_LENGTH),
-			DataTypes.VARBINARY(VarBinaryType.MAX_LENGTH)
+				DataTypes.BINARY(BinaryType.MAX_LENGTH),
```
Comment (bowenli86) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/table/catalog/hive/HiveCatalogDataTypeTest.java]: pls revert unnecessary change

**Comment by zjuwangg** [comment]
Thanks for your detailed review @bowenli86 
I address the comments again, please have a look when you have time.

**Comment by zjuwangg** [comment]
@wuchong 
The comments have all been addressed and Travis test also passed after many times attempt.
I think this pr is ready to merge.
It'll be will gratefull if you can have a look and help merge it when you have time.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveTypeUtil.java [hunk 9]
```diff
@@ -85,65 +86,71 @@ public static TypeInfo toHiveTypeInfo(DataType dataType) {
 		LogicalTypeRoot type = dataType.getLogicalType().getTypeRoot();
 
 		if (dataType instanceof AtomicDataType) {
```
Comment (twalthr) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveTypeUtil.java]: How about using a `LogicalTypeVisitor` or `LogicalTypeDefaultVisitor` here instead? It would simplify the logic here because `if/else/instanceof` checks would not be required anymore. Furthermore, we would force implementers to make sure all cases are covered.

File: docs/dev/table/catalog.md [hunk 10]
```diff
@@ -161,8 +161,8 @@ Currently `HiveCatalog` supports most Flink data types with the following mappin
 | VARCHAR(p)    |  varchar(p)** |
```
Comment (twalthr) [inline docs/dev/table/catalog.md]: nit: make all Hive data types upper case, this is also how the Hive documentation explains them: https://cwiki.apache.org/confluence/display/Hive/LanguageManual+Types#LanguageManualTypes-NumericTypes

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveTypeUtil.java (hunk 9 above)
Comment (zjuwangg) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveTypeUtil.java]: Yep, that would be more elegant.

**Comment by zjuwangg** [comment]
Hi, @twalthr ~
Thanks for your detailed review and valuable comments.
I have just updated my code and please have a look when you have time.

**Comment by zjuwangg** [comment]
Hi, @twalthr ~
Thanks for your detailed review and valuable comments.
I have just updated my code and please have a look when you have time.

File: docs/dev/table/catalog.md [hunk 11]
```diff
@@ -157,28 +157,28 @@ Currently `HiveCatalog` supports most Flink data types with the following mappin
...
+| BOOLEAN       |  BOOLEAN |
+| TINYINT       |  TINYINT |
+| SMALLINT      |  SMALLINT |
+| INT           |  INT |
+| BIGINT        |  LONG |
+| FLOAT         |  FLOAT |
+| DOUBLE        |  DOUBLE |
+| DECIMAL(p, s) |  DECIMAL(p, s) |
+| DATE          |  DATE |
+| TIMESTAMP_WITHOUT_TIME_ZONE |  timestamp |
```
Comment (twalthr) [inline docs/dev/table/catalog.md]: nit: timestamp is still lower case

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveTypeUtil.java [hunk 12]
```diff
@@ -83,112 +102,15 @@ public static String toHiveTypeName(DataType type) {
...
-
-			Map<String, DataType> nameToType = fieldsDataType.getFieldDataTypes();
-			List<TypeInfo> typeInfos = new ArrayList<>(names.size());
-
-			for (String name : names) {
-				typeInfos.add(toHiveTypeInfo(nameToType.get(name)));
-			}
-
-			return TypeInfoFactory.getStructTypeInfo(names, typeInfos);
+//		LogicalTypeRoot type = dataType.getLogicalType().getTypeRoot();
```
Comment (twalthr) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveTypeUtil.java]: remove this line?

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveTypeUtil.java [hunk 13]
```diff
@@ -276,4 +198,202 @@ private static DataType toFlinkPrimitiveType(PrimitiveTypeInfo hiveType) {
...
+					typeInfos.add(typeInfo);
+				} else {
+					return null;
+				}
+			}
+			return TypeInfoFactory.getStructTypeInfo(names, typeInfos);
+		}
+
+		@Override
+		public TypeInfo visit(DistinctType distinctType) {
```
Comment (twalthr) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveTypeUtil.java]: Some last suggestion: Instead of returning null a dozen of times you could simply use the provided `LogicalTypeDefaultVisitor` and throw the exception in the `defaultMethod` there.

Comment (zjuwangg) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveTypeUtil.java]: That would be really cool!
Thanks for your feedback.

File: docs/dev/table/catalog.md [hunk 14]
```diff
@@ -189,11 +190,14 @@ The following limitations in Hive's data types impact the mapping between Flink
```
Comment (bowenli86) [inline docs/dev/table/catalog.md]: should be "*minimum* length is 255". It was my mistake, but would be great to fix that too

**Comment by bowenli86** [comment]
LGTM, @twalthr can you please help to merge it? I'm on paternity leave now. Thanks!

File: docs/dev/table/catalog.md (hunk 14 above)
Comment (zjuwangg) [reply docs/dev/table/catalog.md]: It' not a mistake according to [hive char type](https://cwiki.apache.org/confluence/display/Hive/LanguageManual+Types#LanguageManualTypes-char) definition.
